Sming/system/m_printf.cpp

```diff
@@ -0,0 +1,179 @@
+/*
+Author: (github.com/)ADiea
+Project: Sming for ESP8266 - https://github.com/anakod/Sming
+License: MIT
+Date: 20.07.2015
+Descr: embedded very simple version of printf with float support
+*/
+
+#include <stdarg.h>
+#include "../../SmingCore/SmingCore.h"
+
+#define MPRINTF_BUF_SIZE 256
+
+void (*cbc_printchar)(char ch) = uart_tx_one_char;
+
+#define SIGN    	(1<<1)	/* Unsigned/signed long */
+
+int m_vsnprintf(char *buf, uint32_t maxLen, const char *fmt, va_list args);
+
+void setMPrintfPrinterCbc(void (*callback)(char))
+{
+	cbc_printchar = callback;
+}
+
+int m_printf(const char *fmt, ...)
+{
+	char buf[MPRINTF_BUF_SIZE], *p;
+	va_list args;
+	int n = 0;
+
+	va_start(args, fmt);
+	m_vsnprintf(buf, sizeof(buf), fmt, args);
+	va_end(args);
+
+	p = buf;
+	while (*p)
+	{
+		cbc_printchar(*p);
+		n++;
+		p++;
+	}
+
+	return n;
+}
+
+int m_vsnprintf(char *buf, uint32_t maxLen, const char *fmt, va_list args)
+{
+	int i, base, flags, qualifier;
+	char *str;
+	const char *s;
+	uint8_t precision = 6;
+
+	const uint8_t overflowGuard = 24;
+
+	char tempNum[24];
+
+	for (str = buf; *fmt; fmt++)
+	{
+		if(maxLen - (str - buf) < overflowGuard)
+		{
+			*str++ = '.';
+			*str++ = '.';
+			*str++ = '.';
+			break;
+		}
+
+		if (*fmt != '%')
+		{
+			*str++ = *fmt;
+			continue;
+		}
+
+		flags = 0;
+		fmt++; // This skips first '%'
+
+		do
+		{
+			//skip width and flags data - not supported
```

Comment:
Theese are my test cases, I will probably include some of them in the basic console example 

```
	m_printf( "pi=%f\n", 3.14f);
	m_printf( "1.002=%f\n", 1.002);
	m_printf( "1.050250=%f\n", 1.050250);
	m_printf( "pi=%d %f %c\n", 3, 3.14, '3');
	m_printf( "etc %d %f %s %f %f\n", -99, -3.01040, "abc", 3.0/4, 8.0/9);

	m_printf( "etc %3.0d %.3f %s %.5f %f\n", -99, -3.01040, "abc", 3.0/4, 8.0/9);

	m_printf( "etc %3.0d %% %.3f %%%s %-+.5f %#*f %%%%\n", -99, -3.01040, "abc", 3.0/4, 8.0/9);

	m_printf( "etc %3.0d % %.3f %j %s %-+.5f %#*f %z%k\n", -99, -3.01040, "abc", 3.0/4, 8.0/9);

	int min_i =
	m_printf("Int limits %d .. %d Uint limits 0 .. %u Long limits %ld .. %ld  \n",
			INT_MIN, INT_MAX, UINT_MAX, LONG_MIN, LONG_MAX);

	 m_printf("Color %s, number1 %d, number2 %05d, hex %#x, float %5.2f, unsigned value %u, octal %o.\n",
	        "red", 123456, 89, 0x00C0FFEE, 3.14159, 250, 06675);
```